File: java/memo/CanSum.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

public class CanSum {
 
    private int targetSum;
    private List<Integer> numbers = new LinkedList<>();
    private Map<Integer, Boolean> memo = new HashMap<>();

    public CanSum(int targetSum, List<Integer> numbers) {
        this.targetSum = targetSum;
        this.numbers = numbers;
    }

    private boolean canSum(int targetSum, List<Integer> numbers) {
        if (targetSum == 0) return true;
        if (targetSum < 0) return false;

        for (int num : numbers) {
            int remainder = targetSum - num;

            if (canSum(remainder, numbers)) {
                return true;
            }
        }
        return false;
    }

    private boolean canSumMemo(int targetSum, List<Integer> numbers, Map<Integer, Boolean> memo) {
        if (memo.containsKey(targetSum)) return memo.get(targetSum);
        if (targetSum == 0) return true;
        if (targetSum < 0) return false;

        for (int num : numbers) {
            int remainder = targetSum - num;
            if (canSumMemo(remainder, numbers, memo)) {
                memo.put(targetSum, true);
                return true;
            }
        }
        memo.put(targetSum, false);
        return false;
    }

    
    public void clear() {
        this.numbers.clear();
        this.targetSum = 0;
    }

    public void runRecursive() {
        System.out.println(canSum(this.targetSum, this.numbers));
    }

    public void runMemoRecursive() {
        System.out.println(canSumMemo(this.targetSum, this.numbers, this.memo));
    }
    
    public static void main(String[] args) {
        CanSum c = new CanSum(300, Arrays.asList(7,14,28));
        // c.runRecursive();
        c.runMemoRecursive();
    }
}

File: javascript/memo/canSum.js
/**
 * Write a function `canSum(targetSum, numbers)` that takes in a targetSum
 * and an array of numbers as arguments
 * 
 * The function should return a boolean indicating whether or not it is 
 * possible to generate the targetSum using numbers from hte array.
 * 
 * You may use an elemet of the array as many times as needed.
 * 
 * You may assume that all input numbers are nonnegative.
 */



const canSum = (targetSum, numbers) => {
    if (targetSum === 0) return true;
    if (targetSum < 0) return false;

    for (let num of numbers) {
        const remainder = targetSum - num;

        //* if entering in here, which also is a call to a recursion
        //* the return inside the block basically means 
        //* base case of 0 has been hit and bubbles back up
        if (canSum(remainder, numbers) === true) {
            return true;
        }
    }

    return false;
};


const canSumMemo = (targetSum, numbers, memo={}) => {
    if (targetSum in memo) return memo[targetSum];
    if (targetSum === 0) return true;
    if (targetSum < 0) return false;

    for (let num of numbers) {
        const remainder = targetSum - num;
        if (canSumMemo(remainder, numbers, memo) === true) {
            memo[targetSum] = true;
            return true;
        }
    }
    memo[targetSum] = false;
    return false;
};


console.log(canSum(7, [2,3])); // true
console.log(canSum(7, [5,3,4,7])); // true
console.log(canSum(7, [2,4])); // false
console.log(canSum(8, [2,3,5])); // true


console.log(canSumMemo(300, [7,14])); // false